portfolio/src/main/java/com/google/sps/servlets/DeleteData.java

```diff
@@ -0,0 +1,39 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+/**
+ * Delete data from datastore object when called. 
+ */
+@WebServlet("/deletedata")
+public class DeleteData extends HttpServlet {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    DatastoreService storeData = DatastoreServiceFactory.getDatastoreService();
+    // TODO(morleyd): See if this code generalizes for >500 entries, think there
+    // might be an upper limit for number of entries deleted at a time
+
+    // TODO(morleyd): See if there is a better way to mass delete values, this
+    // seems like a common task and there is likely some optimization to do it
+    // faster
```

Comment:
There are delete overloads that take multiple keys, which would let you do it in one transaction instead of many.

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService#method-summary

Reply:
I was uncertain how to get my input in the proper input for one of the delete overloads. I wanted to use delete(java.lang.Iterable<Key> keys) but I could figure out a good way to construct this object. Intuitively I want to use like the functional programming map and just apply .entity() to each element, but I was unsure of a way to do this in Java and the examples I found used .stream() which doesn't work on an iterable type. I think I'm also somewhat unclear on what the result is before one calls asIterable() so that might explain why I might be overlooking some other clear delete overload if you see one that seems more appropriate.

Reply:
Should I just loop through all the elements and create a new data structure this way? This also seemed horribly inefficient so I'm not sure this is a better solution, though it prevents the partial deleted issue you mentioned.

Reply:
Also it looks like non of the delete methods handle lock the database and instead they throw a java.util.ConcurrentModificationException exception. What would it make sense for one to do in this case to have the desired thread safe beahvior?

Reply:
Database interactions are a non-trivial design, and it will depend heavily on your use patterns.

Right now your only operations are:
A. Add a single comment
B. Delete all comments

A is write, B is read-write. B could involve a lot of rows (in fact, I would normally look for a built in clear operation, but I don't see one for this DB). You want to avoid holding the write lock (so to speak; this would be internally in a DB transaction, not as a Java lock variable) for a long time. But, B has the advantage that each row involved is independent from the others.

So, my first thought for how to write delete would be something like:
1. Retrieve all keys in one transaction.
2. If there are a lot of keys (the actual threshold would depend on profiling), divide them into batches.
3. For each batch, delete that batch in a transaction, ignoring rows that don't exist.

With other clients involved, you can have some inconsistencies with this model. Someone might add a comment between 1 and 3, which you won't delete. Probably okay (the new comment could just as easily have arrived right _after_ 3). Or someone might delete row between 1 and 3. You were already going to delete everything, so that's okay, just make sure the missing rows don't disrupt your transaction.

This analysis would change if you had different operations involved, and depending on the patterns of how common each operation is.

Reply:
Regarding `delete`: Iterable is an interface, so most of your typical collection data structures should work. I think it's fine to construct something in memory in this case, though you're right to be concerned that it could get bad if you had many keys. You could alleviate that with batching.